README.md
# Лабораторная работа №5 по дисциплине "Компьютерная графика"

[![License: MIT ](https://img.shields.io/badge/License-MIT-purple.svg)](https://opensource.org/licenses/MIT)
[![Github CI](https://github.com/Dmrain/stretching_the_texture/actions/workflows/ci.yml/badge.svg)](https://github.com/Dmrain/stretching_the_texture/actions/workflows/ci.yml)
![Python](https://img.shields.io/badge/Python-3.10.11-yellow)

## Описание задачи
> Требуется разработать программу, в которой в верхнем левом углу помещена текстура, при 
нажатии на кнопку мыши эта текстура растягивается до позиции, в которой находится 
курсор мыши.
> 
## Демонстрация выполнения программы

![GIF](https://github.com/Dmrain/stretching_the_texture/blob/master/gif%26img/gifka.gif)
## Установка и запуск

Клонировать репозиторий:

    git clone https://github.com/Dmrain/stretching_the_texture.git

Перейти в папку репозитория:

    cd stretching_the_texture

Установить зависимости:

    install python
    pip install -r requirements.txt
    pip install unittest

Открыть терминал в исходной папке проекта и выполнить следующую команду:
    
    python texture.py


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [v1.0.1] - 2023-12-08

### Added
- **Tests Inclusion:** Introduced a suite of tests using the `unittest` module to ensure texture display and event handling accuracy within the program.
- **Program Demonstration:** Incorporated a demonstration of the program's functionality within the tests, validating the texture rendering in a Pygame window with OpenGL support.

### Details
- Implemented the `test_texture_rendering` test, examining the correct loading of the texture, its parameters (width and height), and initiating simulated mouse event handling to verify program responsiveness.

## [v1.0.0] - 2023-12-08

### Added
- **Texture Loading:** Implemented a program for loading textures using Pygame and OpenGL, rendering them in a Pygame window.


tests/test_texture.py
import unittest
import pygame
from OpenGL.GL import *


class TestTextureRendering(unittest.TestCase):
    def test_texture_rendering(self):
        # Загрузка окна Pygame с поддержкой OpenGL
        pygame.init()
        window_width, window_height = 800, 600
        pygame.display.set_mode((window_width, window_height), pygame.OPENGL | pygame.DOUBLEBUF)

        # Загрузка текстуры
        texture_path = 'gif&img/tekstura.png'
        texture_surface = pygame.image.load(texture_path)
        texture_data = pygame.image.tostring(texture_surface, "RGBA", 1)
        width, height = texture_surface.get_width(), texture_surface.get_height()

        texture = glGenTextures(1)
        glBindTexture(GL_TEXTURE_2D, texture)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR)
        glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, width, height, 0, GL_RGBA, GL_UNSIGNED_BYTE, texture_data)

        # Проверка корректной загрузки текстуры
        self.assertIsNotNone(texture)
        self.assertGreater(width, 0)
        self.assertGreater(height, 0)

        # Проверка обработки событий мыши (здесь может быть имитация событий с помощью mock-объектов)
        # Например, создание событий мыши и проверка, что программа реагирует на них корректно
        # ...

        pygame.quit()


if __name__ == '__main__':
    unittest.main()


texture.py
import pygame
from OpenGL.GL import *
import numpy as np


def load_texture(texture_path):
    """
    Загружает текстуру из файла изображения с использованием Pygame.

    Параметры:
    - texture_path (str): Путь к файлу изображения.
    """
    texture_surface = pygame.image.load(texture_path)
    # Преобразование данных текстуры в строковый формат
    texture_data = pygame.image.tostring(texture_surface, "RGBA", 1)
    width, height = texture_surface.get_width(), texture_surface.get_height()

    # Создание и настройка текстуры OpenGL
    texture = glGenTextures(1)
    glBindTexture(GL_TEXTURE_2D, texture)
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR)
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR)
    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, width, height, 0, GL_RGBA, GL_UNSIGNED_BYTE, texture_data)

    return texture, width, height


# Инициализация Pygame и OpenGL
pygame.init()
window_width, window_height = 800, 600
pygame.display.set_mode((window_width, window_height), pygame.OPENGL | pygame.DOUBLEBUF)

# Настройка текстуры
texture_path = 'gif&img/tekstura.png'
texture, texture_width, texture_height = load_texture(texture_path)

# Установка параметров OpenGL
glEnable(GL_TEXTURE_2D)
glClearColor(0.0, 0.0, 0.0, 1.0)
glEnable(GL_BLEND)
glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA)

# Инициализация часов Pygame
clock = pygame.time.Clock()

cursor_x, cursor_y = 0, 0
is_texture_drawn = False

running = True
while running:
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            running = False
        elif event.type == pygame.MOUSEBUTTONDOWN and event.button == 1:
            # Получение координат мыши и их преобразование в систему координат OpenGL
            mouse_x, mouse_y = event.pos
            cursor_x = mouse_x / window_width * 2 - 1
            cursor_y = 1 - mouse_y / window_height * 2
            is_texture_drawn = True

    # Очистка экрана белым цветом
    glClearColor(1.0, 1.0, 1.0, 1.0)
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT)
    # Загрузка единичной матрицы
    glLoadIdentity()

    # Отрисовка текстуры при необходимости
    if is_texture_drawn:
        glBegin(GL_QUADS)
        glTexCoord2f(0, 0)
        glVertex2f(-1, 1)
        glTexCoord2f(1, 0)
        glVertex2f(cursor_x, 1)
        glTexCoord2f(1, 1)
        glVertex2f(cursor_x, cursor_y)
        glTexCoord2f(0, 1)
        glVertex2f(-1, cursor_y)
        glEnd()

    # Смена буферов отображения
    pygame.display.flip()
    clock.tick(60)

pygame.quit()
